[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138022
* #138014
* #137722

The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

swolchok committed:
[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend

The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

[ghstack-poisoned]

swolchok committed:
fix lint on "[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend"


The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend"


The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

malfet reviewed (approved):
Looks good to me

malfet reviewed (changes requested):
Those ifdefs were added to suppress warnings, rather than anything else when compiling on MacOS 15. We don't have a proper mechism to compile for newer macos but deploy on an older ones

swolchok commented on aten/src/ATen/native/mps/operations/Normalization.mm:
```diff
@@ -718,10 +718,13 @@ static string get_mem_string(c10::MemoryFormat memory_format) {
-#else
-          rsqrtTensor = [mpsGraph reverseSquareRootWithTensor:varianceEpsTensor name:nil];
-#endif
+          if (@available(macOS 15.0, *)) {
```
> do not work

Can you elaborate? I wrote this diff because I was building on macOS 14 with a macOS-15-capable SDK, and tests were failing because reciprocalSquareRootWithTensor:name: was not available at runtime. This diff fixed that, so I'm unclear on how it doesn't work.

swolchok committed:
use is_macos_13_or_newer instead of @available on "[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend"


The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

[ghstack-poisoned]

swolchok committed:
add -Wdeprecated-declarations suppression per request from @malfet on "[PyTorch] Fix incorrect macOS 15.0 gating in MPS backend"


The ifdef as written just checks if the macOS 15.0-capable SDK is being used. You also need a runtime gate to make sure macOS 15 is in use.

Differential Revision: [D64429453](https://our.internmc.facebook.com/intern/diff/D64429453/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge
